File: Redcoder/Python/33.py
from collections import deque
H, W = map(int,input().split())
box = [[-1] * (W+2) for _ in range(H+2)]
num = []
for i in range(H):
    a = list(input())
    num.append(a)

count = 0
for i in range(H):
    for j in range(W):
        if num[i][j] == '.':
            box[i+1][j+1] = 0
            count += 1


Q = deque()
Q.append([1,1,1])
box[1][1] = 1
X = [0,-1,0,1]
Y = [-1,0,1,0]
ans = 10**9+7
while len(Q) > 0:
    y,x,cost = Q.popleft()
    for i in range(4):
        yy = y + Y[i]
        xx = x + X[i]
        if yy == H and xx == W:
            ans = min(ans, cost+1)
        elif box[yy][xx] == 0:
            Q.append([yy,xx,cost+1])
            box[yy][xx] = cost
if ans == 10**9+7:
    print(-1)
else:
    print(count - ans)

File: Redcoder/C++/33.cpp
#include <bits/stdc++.h>
#define rep(i,nn,n) for (int i = nn; i < (n); ++i)
#define rrep(i,nn,n) for (int i = nn; i > (n); --i)
using namespace std;
using ll = long long;
using P = pair<int,int>;
#define chmin(x,y) x = min(x,y)
#define chmax(x,y) x = max(x,y)

const int MAX_NUM = 55;
int H,W;
vector<vector<int>> box(MAX_NUM, vector<int>(MAX_NUM, -1));

int BFS(){
    int dx[4] = {0,-1,0,1};
    int dy[4] = {-1,0,1,0};
    queue<P> q;
    q.push({1,1});
    box[1][1] = 1;
    while (!q.empty()){
        P pos = q.front();q.pop();
        rep(i,0,4){
            int y = dy[i] + pos.first;
            int x = dx[i] + pos.second;
            if (box[y][x] == 0){
                box[y][x] = 1 + box[pos.first][pos.second];
                q.push({y,x});
            }
        }
    }
    return box[H][W];
}

int main() {
    int ans;
    cin >> H >> W;
    rep(i,0,H){
        string s;
        cin >> s;
        rep(j,0,W){
            if (s[j] == '.'){
                box[i+1][j+1] = 0;
                ++ans;
            }
        }
    }

    int count = BFS();
    if (count == -1){
        cout << count << endl;
        return 0;
    }
    ans -= count;
    cout << ans << endl;
    return 0;
}